### PR TITLE
add action mailer instrumentation

### PR DESCRIPTION
**Pull request by ericmustin**
This PR provides instrumentation support for Rails 5/6 ActionMailer.

This PR addresses #250 , instruments the `process.action_mailer` and `deliver.action_mailer` ActiveSupport::Notification Events for the [Rails ActionMailer](https://guides.rubyonrails.org/action_mailer_basics.html) Module.

There's perhaps some additional metadata that could be collected as span tags but a lot of it (`from` , `to` `subject` `body` etc etc) seems like it would contain PII, so i've limited the tags to what's useful for measuring execution times and errors between specific mailers and actions.

I will try to update this PR shortly with an example trace for a small sample application

### CODE REVIEW

File: lib/ddtrace/contrib/action_mailer/events/process.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+          def event_name
+            self::EVENT_NAME
+          end
+
+          def span_name
+            Ext::SPAN_PROCESS
+          end
+
+          def span_type
+            # ActionMailer creates emails like a controller
```
Comment (marcotc) [inline lib/ddtrace/contrib/action_mailer/events/process.rb:25]: Could you clarify this? Are these top-level spans then? Do they happen on the same execution context as, say, a Rack request?

File: spec/ddtrace/contrib/action_mailer/helpers.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+require 'rails'
+
+RSpec.shared_context 'ActionMailer helpers' do
+  before(:each) do
+    if ActionMailer::Base.respond_to?(:delivery_method)
```
Comment (marcotc) [inline spec/ddtrace/contrib/action_mailer/helpers.rb:5]: Is this condition for? Different version of Rails?

**Review by marcotc — COMMENTED** [review]
Looks very good! I asked a few clarification questions, but very minor ones.

File: lib/ddtrace/contrib/action_mailer/events/process.rb (hunk 1 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/action_mailer/events/process.rb:25]: i believe they are _not_ top level spans (although, if executed from say, the rails console i suppose they could be) but i am going to try to stress test this in a sample app, will report back when i have more. tl;dr i believe i've botched the span_type here and need to update, but want to confirm that

Comment (ericmustin) [reply lib/ddtrace/contrib/action_mailer/events/process.rb]: So, to update: So this integration is similar to active record or action view in that there's a few different spans which could all be "top level" in the sense that they're the `1st` span for the service, but generally speaking they'll rarely be the _root_ span of the trace, usualy they're a child of action_pack controller span or a job consumer (delayed job, sidekiq, etc) span.

Here's a few examples:

1. async job queue (deliver_later)

Image 2021-07-29 at 4 01 30 PM

2. syncronous, no job queue (deliver_now)

Image 2021-07-29 at 1 51 29 PM

I've tried to update the span_types to reflect existing standards that are closer to their more appropriate usage, marking them as either a `template` (which is what `.process` spans do, they render the mailer templates) or `worker` (which is what `.deliver` spans do, they send emails).

Anyway lmk what you think

File: spec/ddtrace/contrib/action_mailer/helpers.rb (hunk 2 above)
Comment (ericmustin) [reply spec/ddtrace/contrib/action_mailer/helpers.rb:11]: added a comment

File: lib/ddtrace/contrib/action_mailer/events/process.rb (hunk 1 above)
Comment (marcotc) [reply lib/ddtrace/contrib/action_mailer/events/process.rb]: Makes sense. I like the traces you showed above, they all seem reasonable to me 👍.

**Comment by marcotc** [comment]
Are the information about the email being sent (from, to, subject, payload size) not available in the ActiveSupport notification?
Even if it contains PII, it would be good to have it optionally captured (possibly behind a `c.use :action_mailer, email_data: true` flag).

**Comment by marcotc** [comment]
The actual network operations also seem to be missing, but I think that the current instrumentation level provides enough value as it.
We can add network-level `Net::SMTP` instrumentation separately later on.

**Comment by ericmustin** [comment]
@marcotc lmk what you think, sorry didnt realize this was still open

File: lib/ddtrace/contrib/action_mailer/events/deliver.rb [hunk 3]
```diff
@@ -32,6 +32,19 @@ def process(span, event, _id, payload)
             span.span_type = span_type
             span.set_tag(Ext::TAG_MAILER, payload[:mailer])
             span.set_tag(Ext::TAG_MSG_ID, payload[:message_id])
+
+            # Since email date can contain PII we disable by default
+            # Some of these fields can be either strings or arrays, so we try to normalize
```
Comment (marcotc) [inline lib/ddtrace/contrib/action_mailer/events/deliver.rb:37]: 👍

**Review by marcotc — APPROVED** [review]
🛤️🛫

**Comment by hale** [comment]
I figure you folks have this covered but just in case not, this isn't yet in the changelog for 0.53

c.f. https://github.com/DataDog/dd-trace-rb/pull/1711/files

**Pull request merged**